File: Samples/Geolocation/js/geolocationjs/js/scenario6_GetLastVisit.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved
//// This code is licensed under the MIT License (MIT).

//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved
//// This code is licensed under the MIT License (MIT).

(function () {
    "use strict";

    // Namespace abbreviations.
    var Geolocation = Windows.Devices.Geolocation;

    var getLastVisitButton;
    var latitudeElement;
    var longitudeElement;
    var accuracyElement;
    var timestampElement;
    var statechangeElement;

    var page = WinJS.UI.Pages.define("/html/scenario6_GetLastVisit.html", {
        ready: function (element, options) {
            getLastVisitButton = document.getElementById("getLastVisitButton");
            latitudeElement = document.getElementById("latitude");
            longitudeElement = document.getElementById("longitude");
            accuracyElement = document.getElementById("accuracy");
            timestampElement = document.getElementById("timestamp");
            statechangeElement = document.getElementById("statechange");

            getLastVisitButton.addEventListener("click", getLastVisit, false);
            getLastVisitButton.disabled = false;
        }
    });

    function getLastVisit() {
        getLastVisitButton.disabled = true;

        Geolocation.Geolocator.requestAccessAsync().then(function (accessStatus) {
            switch (accessStatus) {
                case Geolocation.GeolocationAccessStatus.allowed:

                    WinJS.log && WinJS.log("Waiting for update...", "sample", "status");
                    getLastVisitButton.disabled = true;

                    // Get the last visit report, if any. This should be a quick operation
                    return Geolocation.GeovisitMonitor.getLastReportAsync().then(function (visit) {
                        WinJS.log && WinJS.log("Visit info updated", "sample", "status");
                        return visit;
                    });

                case Geolocation.GeolocationAccessStatus.denied:
                    WinJS.log && WinJS.log("Access to location is denied.", "sample", "error");
                    break;

                case Geolocation.GeolocationAccessStatus.unspecified:
                    WinJS.log && WinJS.log("Unspecified error!", "sample", "error");
                    break;
            }
        }).done(function (visit) {
            updateLastVisit(visit);
            getLastVisitButton.disabled = false;
        });
    }

    function updateLastVisit(visit) {
        if (!visit) {
            latitudeElement.innerHTML = "No data";
            longitudeElement.innerHTML = "No data";
            accuracyElement.innerHTML = "No data";
            timestampElement.innerHTML = "No data";
            statechangeElement.innerHTML = "No data";
        } else {
            statechangeElement.innerHTML = SdkSample.toStateChangeString(visit.stateChange);

            var dateTimeFormatter = new Windows.Globalization.DateTimeFormatting.DateTimeFormatter("shortdate longtime");
            timestampElement.innerHTML = dateTimeFormatter.format(visit.timestamp);

            // Check if there is a valid position in the report.
            if (visit.position === null) {
                latitudeElement.innerHTML = "No data";
                longitudeElement.innerHTML = "No data";
                accuracyElement.innerHTML = "No data";
            } else {
                var coord = visit.position.coordinate;
                latitudeElement.innerHTML = coord.point.position.latitude;
                longitudeElement.innerHTML = coord.point.position.longitude;
                accuracyElement.innerHTML = coord.accuracy;
            }
        }
    }

})();

File: Samples/Geolocation/cppwinrt/Scenario6_GetLastVisit.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario6_GetLastVisit.h"
#include "Scenario6_GetLastVisit.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace winrt::Windows::Devices::Geolocation;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;

namespace winrt::SDKTemplate::implementation
{
    Scenario6_GetLastVisit::Scenario6_GetLastVisit()
    {
        InitializeComponent();
    }

    fire_and_forget Scenario6_GetLastVisit::GetLastVisitButtonClicked(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        GetLastVisitButton().IsEnabled(false);
        LocationDisabledMessage().Visibility(Visibility::Collapsed);

        // Request permission to access location
        auto accessStatus = co_await Geolocator::RequestAccessAsync();

        Geovisit visit = nullptr;
        switch (accessStatus)
        {
        case GeolocationAccessStatus::Allowed:
            m_rootPage.NotifyUser(L"Waiting for update...", NotifyType::StatusMessage);

            // Get the last visit report, if any. This should be a quick operation.
            visit = co_await GeovisitMonitor::GetLastReportAsync();
            m_rootPage.NotifyUser(L"Visit info updated.", NotifyType::StatusMessage);
            break;

        case GeolocationAccessStatus::Denied:
            m_rootPage.NotifyUser(L"Access to location is denied.", NotifyType::ErrorMessage);
            LocationDisabledMessage().Visibility(Visibility::Visible);
            break;

        default:
        case GeolocationAccessStatus::Unspecified:
            m_rootPage.NotifyUser(L"Unspecified error.", NotifyType::ErrorMessage);
            break;
        }

        UpdateLastVisit(visit);

        GetLastVisitButton().IsEnabled(true);
    }

    // Updates the user interface with the Geovisit data provided
    void Scenario6_GetLastVisit::UpdateLastVisit(Geovisit const& visit)
    {
        if (visit == nullptr)
        {
            ScenarioOutput_Latitude().Text(L"No data");
            ScenarioOutput_Longitude().Text(L"No data");
            ScenarioOutput_Accuracy().Text(L"No data");
            ScenarioOutput_Timestamp().Text(L"No data");
            ScenarioOutput_VisitStateChange().Text(L"No data");
        }
        else
        {
            // A valid visit is available, extract the state change and Timestamp.
            ScenarioOutput_VisitStateChange().Text(to_hstring(visit.StateChange()));
            ScenarioOutput_Timestamp().Text(FormatDateTime(visit.Timestamp(), L"shortdate shorttime"));

            // If a valid position is available, extract the position information that caused the state change to happen.
            if (visit.Position() == nullptr)
            {
                ScenarioOutput_Latitude().Text(L"No data");
                ScenarioOutput_Longitude().Text(L"No data");
                ScenarioOutput_Accuracy().Text(L"No data");
            }
            else
            {
                ScenarioOutput_Latitude().Text(to_hstring(visit.Position().Coordinate().Point().Position().Latitude));
                ScenarioOutput_Longitude().Text(to_hstring(visit.Position().Coordinate().Point().Position().Longitude));
                ScenarioOutput_Accuracy().Text(to_hstring(visit.Position().Coordinate().Accuracy()));
            }
        }
    }
}
